tensorboard/data/server/run.rs

```diff
@@ -0,0 +1,363 @@
+        let f2_name = logdir.path().join("tfevents.456");
+        let mut f1 = BufWriter::new(File::create(&f1_name)?);
+        let mut f2 = BufWriter::new(File::create(&f2_name)?);
+
+        // Write file versions.
+        for (f, wall_time) in &mut [(&mut f1, 1234.0), (&mut f2, 2345.0)] {
+            let file_version = pb::Event {
+                wall_time: *wall_time,
+                what: Some(pb::event::What::FileVersion("brain.Event:2".to_string())),
+                ..Default::default()
+            };
+            write_event(f, &file_version)?;
+        }
+
+        // Write some data points across both files.
+        let tag = Tag("accuracy".to_string());
+        write_scalar(&mut f1, &tag, Step(0), WallTime::new(1235.0).unwrap(), 0.25)?;
+        write_scalar(&mut f1, &tag, Step(1), WallTime::new(1236.0).unwrap(), 0.50)?;
+        write_scalar(&mut f2, &tag, Step(2), WallTime::new(2346.0).unwrap(), 0.75)?;
+        write_scalar(&mut f2, &tag, Step(3), WallTime::new(2347.0).unwrap(), 1.00)?;
+        f1.into_inner()?.sync_all()?;
+        f2.into_inner()?.sync_all()?;
+
+        let mut loader = RunLoader::new();
+        loader.reload(vec![f1_name, f2_name]);
+
+        // Start time should be that of the file version event, even though that didn't correspond
+        // to any time series.
+        assert_eq!(loader.start_time, Some(WallTime::new(1234.0).unwrap()));
+        assert_eq!(loader.time_series.keys().collect::<Vec<_>>(), vec![&tag]);
+        let ts = loader.time_series.get_mut(&tag).unwrap();
+        assert_eq!(
+            *ts.metadata,
+            pb::SummaryMetadata {
+                plugin_data: Some(pb::summary_metadata::PluginData {
+                    plugin_name: crate::data_compat::SCALARS_PLUGIN_NAME.to_string(),
+                    ..Default::default()
+                }),
+                data_class: pb::DataClass::Scalar.into(),
+                ..Default::default()
+            }
+        );
+        let mut basin = crate::reservoir::Basin::new();
+        ts.rsv.commit(&mut basin);
+
+        // Points should be as expected (no downsampling at these sizes).
+        let mut actual_points = Vec::new();
+        for (step, StageValue { wall_time, payload }) in basin.as_slice() {
+            if let StagePayload::Summary(value_box) = payload {
+                if let pb::summary::value::Value::SimpleValue(f) = **value_box {
+                    actual_points.push((*step, *wall_time, f));
+                    continue;
+                }
+            }
+            panic!("step {:?}: {:?}", step, payload);
+        }
+        assert_eq!(
+            actual_points,
+            vec![
+                (Step(0), WallTime::new(1235.0).unwrap(), 0.25),
+                (Step(1), WallTime::new(1236.0).unwrap(), 0.50),
+                (Step(2), WallTime::new(2346.0).unwrap(), 0.75),
+                (Step(3), WallTime::new(2347.0).unwrap(), 1.00),
```

Comment:
Testing my understanding: I assume the `loader.time_series` does not actually sort/order events by step, right? It is just happens to be sorted since we wrote as such in L311-L314. Don't think I've read the reservoir module thoroughly to know that on top of my head, but it looks like `rsv.offer` simply pushes onto the vec.

Reply:
It is sorted because the reservoir preempts any non-preceding records
*before* pushing onto the vector, so at the time that it pushes the step
of the new record is always larger than the steps of all other records.

Including a preemption in this test is probably a pretty reasonable
idea. I’ll go ahead and do that.